sklearn/neighbors/approximate.py

```diff
@@ -226,10 +226,15 @@ def _compute_distances(self, query, candidates):
             # needed since _fit_X[np.array([])] doesn't work if _fit_X sparse
             return np.empty(0, dtype=np.int), np.empty(0, dtype=float)
 
-        distances = pairwise_distances(query, self._fit_X[candidates],
+        if sparse.issparse(self._fit_X):
+            candidate_X = self._fit_X[candidates]
+        else:
+            candidate_X = self._fit_X.take(candidates, axis=0, mode='clip')
+        distances = pairwise_distances(query, candidate_X,
                                        metric='cosine')[0]
         distance_positions = np.argsort(distances)
-        return distance_positions, distances[distance_positions]
+        distances = distances.take(distance_positions, mode='clip', axis=0)
```

Comment:
.take (at least last time I checked which was a while ago) was faster than fancy indexing pretty much always. I am curious about the mode='clip' part - I thought that was only useful if you pass in a return array as out= (though I guess we are using distances as out so maybe it ends up being the same)